packages/meteor/dynamics_nodejs.js

```diff
@@ -1,120 +1,103 @@
+   * @method get
+   * @locus Anywhere
+   * @returns {any} The current value of the variable, or `undefined` if no
+   */
+  get() {
+    if (this.slot !== Meteor._getValueFromAslStore(SLOT_CALL_KEY)) {
+      return;
+    }
+    return Meteor._getValueFromAslStore(CURRENT_VALUE_KEY_NAME);
+  }
 
-// Most Meteor code ought to run inside a fiber, and the
-// _nodeCodeMustBeInFiber assertion helps you remember to include appropriate
-// bindEnvironment calls (which will get you the *right value* for your
-// environment variables, on the server).
-//
-// In some very special cases, it's more important to run Meteor code on the
-// server in non-Fiber contexts rather than to strongly enforce the safeguard
-// against forgetting to use bindEnvironment. For example, using `check` in
-// some top-level constructs like connect handlers without needing unnecessary
-// Fibers on every request is more important that possibly failing to find the
-// correct argumentChecker. So this function is just like get(), but it
-// returns null rather than throwing when called from outside a Fiber. (On the
-// client, it is identical to get().)
-EVp.getOrNullIfOutsideFiber = function () {
-  if (!Fiber.current)
-    return null;
-  return this.get();
-};
+  getOrNullIfOutsideFiber() {
+    return this.get();
+  }
 
-/**
- * @summary Set the environment variable to the given value while a function is run
- * @locus Anywhere
- * @method withValue
- * @memberof Meteor.EnvironmentVariable
- * @param {Any} value Value the environment variable should be set to
- * @param {Function} func The function to run
- * @return {Any} Return value of function
- */
-EVp.withValue = function (value, func) {
-  Meteor._nodeCodeMustBeInFiber();
-
-  if (!Fiber.current._meteor_dynamics)
-    Fiber.current._meteor_dynamics = [];
-  var currentValues = Fiber.current._meteor_dynamics;
-
-  var saved = currentValues[this.slot];
-  try {
-    currentValues[this.slot] = value;
-    return Meteor.wrapFn(func)();
-  } finally {
-    currentValues[this.slot] = saved;
+  /**
+   * @summary takes a value and a function, calls the function with the value set for the duration of the call
+   * @memberof Meteor.EnvironmentVariable
+   * @method withValue
+   * @param {any} value The value to set for the duration of the function call
+   * @param {Function} func The function to call with the new value of the
+   * @param {Object} [options] Optional additional options
+   * @returns {Promise<any>} The return value of the function
+   */
+  withValue(value, func, options = {}) {
+    return Meteor._runAsync(
+      async () => {
+        let ret;
+        try {
+          Meteor._updateAslStore(CURRENT_VALUE_KEY_NAME, value);
+          ret = await func();
+        } finally {
+          Meteor._updateAslStore(CURRENT_VALUE_KEY_NAME, undefined);
+        }
+        return ret;
+      },
+      this,
+      {
+        callId: `${this.slot}-${Math.random()}`,
+        [SLOT_CALL_KEY]: this.slot,
```

Comment:
Does this mean I can't call withValue from inside another withValue call? If I set the same variable, the inner call will clobber the value with undefined, and if I set a different variable, I can't access the originally modified variable because the slot won't match.

Reply:
I wrote a test case for this to confirm: https://github.com/Torgen/meteor/commit/57b314b1a224f41b989a37b3b480e8dcc1824c09

Reply:
Filed https://github.com/meteor/meteor/issues/12702